File: HW2/Problem5B-D/stoogesort.js
"use strict";

const fs = require("fs");

const getRandomInt = (min, max) => {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min)) + min;
};

const genRandomNumbers = numToGen => {
    let numArray = [];

    for(let i = 0; i < numToGen; i++) {
        numArray.push(getRandomInt(0,10000));
    }

    return numArray;
};

const stoogeSort = (array, lowIndex, highIndex) => {
  const n = highIndex - lowIndex + 1;

  if (n === 2 && array[lowIndex] > array[lowIndex + 1]) {
      const storage = array[lowIndex];
      array[lowIndex] = array[lowIndex + 1];
      array[lowIndex + 1] = storage;
  }

  else if (n > 2) {
      const m = Math.ceil((2 * n) / 3);
      stoogeSort(array, lowIndex, m - 1 + lowIndex);
      stoogeSort(array, n - m + lowIndex, highIndex);
      stoogeSort(array, lowIndex, m - 1 + lowIndex);
  }
};

const timeArray = [500,1000,1500,2000,2500,3000,3500];

fs.writeFileSync("stoogetime.csv", "num,time\n");

timeArray.forEach(num => {
    for(let i = 0; i < 5; i++) {
        const randomArray = genRandomNumbers(num);
        const start = new Date();
        const hrstart = process.hrtime();
        stoogeSort(randomArray, 0, randomArray.length - 1);
        const end = new Date() - start;
        const hrend = process.hrtime(hrstart);
        fs.appendFileSync("stoogetime.csv", num + "," + end + "\n");
    }
});

File: HW2/Problem5B-D/stoogesort.py
from random import randint
import timeit

def genRandomNumbers(numToGen):
    numArray = []

    for x in xrange(0,numToGen):
        numArray.append(randint(0,10000))

    return numArray


def stooge_sort(array, low_index, high_index):
    n = high_index - low_index + 1

    if n == 2 and array[low_index] > array[low_index + 1]:
        storage = array[low_index]
        array[low_index] = array[low_index + 1]
        array[low_index + 1] = storage
    elif n > 2:
        m = int(math.ceil(2 * float(n) / 3))
        stooge_sort(array, low_index, m - 1 + low_index)
        stooge_sort(array, n - m + low_index, high_index)
        stooge_sort(array, low_index, m - 1 + low_index)


timeArray = (1000,2000,5000,10000,20000,30000,35000)

outFile = open("stoogetime.csv", "w")

outFile.write("num,time\n")

for num in timeArray:
    for x in xrange(0,5):
        randomArray = genRandomNumbers(num)
        start_time = timeit.default_timer()
        stooge_sort(randomArray)
        elapsed = timeit.default_timer() - start_time

        outFile.write(str(num) + "," + str(elapsed) + '\n')

outFile.close()
